File: Schema/SchemaConfig/tables/Group/CGroup.customisation.vb
Imports System
Imports System.Data
Imports System.Collections.Generic

Imports Framework

'Table-Row Class (Customisable half)
Partial Public Class CGroup

#Region "Constants"
    Public Const [DEFAULT] As String = "Default"
#End Region

#Region "Constructors (Public)"
    'Default Connection String
    Public Sub New()
        MyBase.New()
    End Sub
    
    'Alternative Connection String
    Public Sub New(ByVal dataSrc as CDataSrc)
        MyBase.New(dataSrc)
    End Sub
    
    'Hidden (UI code should use cache instead)
    Protected Friend Sub New(groupId As Integer)
        MyBase.New(groupId)
    End Sub
    Protected Friend Sub New(ByVal dataSrc As CDataSrc, groupId As Integer)
        MyBase.New(dataSrc, groupId)
    End Sub
    Protected Friend Sub New(ByVal dataSrc As CDataSrc, ByVal groupId As Integer, ByVal txOrNull As IDbTransaction)
        MyBase.New(dataSrc, groupId, txOrNull)
    End Sub
#End Region

#Region "Default Values"
    Protected Overrides Sub InitValues_Custom()
        m_groupSortOrder = 0
    End Sub
#End Region

#Region "Default Connection String"
    Protected Overrides Function DefaultDataSrc() As CDataSrc
        Return CDataSrc.Default
    End Function
#End Region

#Region "Properties - Relationships"    
    'Relationships - Foriegn Keys (e.g parent)

    'Relationships - Collections (e.g. children)
    Public ReadOnly Property Settings() As CSettingList
        Get
            Return CSetting.Cache.GetByGroupId(Me.GroupId)
        End Get
    End Property
#End Region

#Region "Properties - Customisation"
    Public ReadOnly Property SettingsCount() As Integer
        Get
            Return Settings.Count
        End Get
    End Property
#End Region

#Region "Save/Delete Overrides"
    Public Overrides Sub Delete(ByVal txOrNull As System.Data.IDbTransaction)
        If IsNothing(txOrNull) AndAlso DataSrc.IsLocal Then
            BulkDelete(Me)
            Exit Sub
        End If

        Settings.DeleteAll(txOrNull)
        MyBase.Delete(txOrNull)
    End Sub
#End Region

#Region "Search Logic"
    'Public Interfaces: Refer to list class for Search (and GetBy*) methods e.g. CGroup.Cache.Search

    'Complex Search Logic (e.g string-based searching) Note: Index-based filters (e.g. fk/bool) are handled in the Search method, and can be ignored here
    Friend Function Match(ByVal filters As CSearchFilters) As Boolean
        With filters
            'Apply active filters here i.e. return false if any filter is active and excludes this record
            If Not String.IsNullOrEmpty(.Name) AndAlso Not Me.GroupName.ToLower.StartsWith(.Name) Then Return False 'Starts-with logic (trailing wildcard)
        End With
        Return True
    End Function
#End Region

#Region "Search Filters"
    'Filters i.e. UI-driven collection of potential search filters (e.g. 'name' from textbox could search firstname/lastname columns)
    Public Class CSearchFilters
        'Constructors (add more for common combinations, or introduce more classes)
        Public Sub New(ByVal name As String)
            Me.Name = name.ToLower()
        End Sub

        'Members (filter info)
        Public Name As String = String.Empty

        'Properties (checks if a complex search is required i.e. some filters are active, other than index-based ones)
        Public Readonly Property NonIndexedFiltersAreNull As Boolean
            Get
                Return True 'e.g. Return String.IsNullOrEmpty(Me.Name) AndAlso ...
            End Get
        End Property
End Class
#End Region

#Region "Caching Details"
    'Cache data
    Private Shared Function LoadCache() As CGroupList
        Return New CGroup().SelectAll()
    End Function
    'Cache Timeout
    Private Shared Sub SetCache(ByVal key As String, ByVal value As CGroupList)
        If Not IsNothing(value) Then value.Sort()
        CCache.Set(key, value) 'Optional parameter can override timeout (default is 3 days)
    End Sub
    'Helper Method
    Private Function CacheGetById(ByVal list As CGroupList) As CGroup
        Return list.GetById(Me.GroupId)
    End Function
#End Region

#Region "ToXml"
    Protected Overrides Sub ToXml_Custom(ByVal w As System.Xml.XmlWriter)
        'Store(w, "Example", Me.Example)
    End Sub
#End Region

End Class

File: Schema/SchemaDeploy/tables/Group/CGroup.customisation.cs
using System;
using System.Text;
using System.Data;
using System.Collections;
using System.Collections.Generic;

using Framework;

namespace SchemaDeploy
{

    //Table-Mapping Class (Customisable half)
    public partial class CGroup
    {
        #region Constants
        #endregion

        #region Constructors (Public)
        //Default Connection String
        public CGroup() : base() {}

        //Alternative Connection String
        public CGroup(CDataSrc dataSrc) : base(dataSrc) {}
        
        //Hidden  (UI code should use cache instead)
        protected internal CGroup(int groupId) : base(groupId) {}
        protected internal CGroup(CDataSrc dataSrc, int groupId) : base(dataSrc, groupId) {}
        protected internal CGroup(CDataSrc dataSrc, int groupId, IDbTransaction txOrNull) : base(dataSrc, groupId, txOrNull) { }
        #endregion

        #region Default Values
        protected override void InitValues_Custom()
        {
            //_sampleDateCreated = DateTime.Now;
            //_sampleSortOrder   = 0;
        }
        #endregion
        
        #region Default Connection String
        protected override CDataSrc DefaultDataSrc()  { return CDataSrc.Default;  }
        #endregion

        #region Properties - Relationships
        //Relationships - Foriegn Keys (e.g parent)

        //Relationships - Collections (e.g. children)
        public CKeyList Keys { get { return CKey.Cache.GetByGroupId(this.GroupId); } }
        protected CKeyList Keys_(IDbTransaction tx) { return new CKey(DataSrc).SelectByGroupId(this.GroupId, tx); } //Only used for cascade deletes
        public int KeyCount {  get { return Keys.Count; } }
        #endregion

        #region Properties - Customisation
        //Derived/ReadOnly (e.g. xml classes, presentation logic)
        public string NameAndCount {  get { return CUtilities.NameAndCount(GroupName, Keys); } }
        #endregion

        #region Save/Delete Overrides
        //Can Override base.Save/Delete (e.g. Cascade deletes, or insert related records)
        #endregion

        #region Custom Database Queries
        //(Not normally required for cached classes, use list class for searching etc)
        //For Stored Procs can use: MakeList (matching schema), or DataSrc.ExecuteDataset (reports etc)
        //For Dynamic sql, can use: SelectSum, SelectDistinct, SelectCount, SelectWhere (inherited methods)
        #endregion

        #region Searching (Optional)
        //For cached classes, custom seach logic resides in static methods on the list class
        // e.g. CGroup.Cache.Search("...")

        //See also the auto-generated methods based on indexes
        //' e.g. CGroup.Cache.GetBy...
        #endregion

        #region Caching Details
        //Cache Key
        internal static string CACHE_KEY = typeof(CGroup).ToString();    //TABLE_NAME

        //Cache data
        private static CGroupList LoadCache()  {   return new CGroup().SelectAll();   }
        //Cache Timeout
        private static void SetCache(CGroupList value)
        {
            if (null != value)  
                value.Sort(); 
            CCache.Set(CACHE_KEY, value);    //Optional parameter can override timeout (otherwise uses config-settings, which default to 3hrs)
        }
        //Helper Method
        private CGroup CacheGetById(CGroupList list)  { return list.GetById(this.GroupId);    }
        #endregion

        #region Cloning
        public CGroup Clone(CDataSrc target, IDbTransaction txOrNull) //, int parentId)
        {
            //Shallow copy: Copies the immediate record, excluding autogenerated Pks
            CGroup copy = new CGroup(this, target);

            //Deep Copy - Child Entities: Cloned children must reference their cloned parent
            //copy.SampleParentId = parentId;

            copy.Save(txOrNull);

            //Deep Copy - Parent Entities: Cloned parents also clone their child collections
            //this.Children.Clone(target, txOrNull, copy.GroupId);

            return copy;
        }
        #endregion

        #region ToXml
        protected override void ToXml_Custom(System.Xml.XmlWriter w)
        {
            //Store(w, "Example", this.Example)
        }
        #endregion
    }
}